### ISSUE TITLE
ttnn.pow(0, NaN) returns 0 on the binary SFPU path; should be NaN (unary path is already correct)

### ISSUE THREAD

**Issue by iklemenskiTT** [tenstorrent/tt-metal#53922]
### Summary

On the binary `pow` SFPU path, a NaN exponent over a zero base returns `+0` instead of NaN. The unary path is unaffected and returns NaN correctly. This was already wrong before #53367 (it returned `inf`), so #53367 did not introduce the wrong answer — but it changed the failure mode from loud to silent, which is worse for the workload the fix targets.

Measured on Blackhole p100a, `ttnn.pow(zeros, exponent)`, output as raw bit patterns:

| exponent | binary, pre-#53367 | binary, current | unary (both) | torch |
|---|---|---|---|---|
| NaN, fp32 | `0x7F800000` inf | **`0x00000000`** | `0x7FC00000` NaN ✓ | NaN |
| NaN, bf16 | `0x7F800000` inf | **`0x00000000`** | `0x7F800000` inf † | NaN |
| `+inf`, fp32 | `0x7F800000` inf | `0x00000000` ✓ | `0x00000000` ✓ | `0` |

† bf16 cannot represent the result: the kernel writes NaN and the packer converts it to inf (tenstorrent/tt-llk#675). Best achievable on bf16 is inf, not NaN.

Note #53367 *fixed* `pow(0, +inf)` as a side effect — it returned `inf` before and now correctly returns `0`.

### Why it matters

Zero bases arise from padding, masked positions, ReLU outputs and freshly-initialised buffers — that is the motivation for #51468. NaN is the other common "not data" sentinel, so a NaN exponent over a zero base is the natural collision of two masked tensors in this op. `inf` propagates and gets noticed; `+0` is indistinguishable from a legitimate result, so the "this lane is invalid" signal is destroyed silently. IEEE-754 specifies NaN in → NaN out precisely to preserve that marker.

Scope is narrow — binary only, zero base only — but it is silent.

### Root cause

SFPU comparisons are bit-pattern tests, not IEEE predicates. Quiet NaN `0x7FC00000` reads as a large positive pattern, so it satisfies both `pow != 0.f` and `pow > 0.f` and falls into the zero fill. This is true of both guard shapes: the `v_and` form that merged and the flat `v_if` pair it replaced produce bit-identical `0x00000000` here.

`ckernel_sfpu_binary_pow.h`, both `_sfpu_binary_power_21f_` and `_sfpu_binary_power_f32_`:

```c
v_if(absbase == 0.f) {
    v_if(pow != 0.f) {          // true for NaN
        y = 0.0f;               // <- NaN lands here
        v_and(pow < 0.f);       // false for NaN
        y = std::numeric_limits<float>::quiet_NaN();
    }
    v_endif;
}
v_endif;
```

### Why it isn't a one-line fix

The awkward value is `+inf`, not NaN. IEEE wants `pow(0, +inf) = +0` but `pow(0, NaN) = NaN`, and NaN and inf share biased exponent `0xFF`, differing only in the mantissa. Every arithmetic shortcut that catches NaN also catches inf:

- `pow - pow` → NaN for NaN, also NaN for inf
- `pow * 0.0f` → NaN for NaN, also NaN for inf
- `v_if(pow != pow)` → **measured**: fires for NaN *and* `±inf`, turning the currently-correct `pow(0, +inf) = 0` into NaN. Do not use this; sfpi's `!=` is not bit-exact and appears to be subtract-then-test.

Separating them requires inspecting the mantissa.

### Candidate fix, with measured cost

Inside the existing `absbase == 0.f` predicate — so no effect on non-zero bases:

```c
// NaN exponent: magnitude bits above +inf's pattern.
v_if(sfpi::as<sfpi::vInt>(sfpi::abs(pow)) > 0x7F800000) {
    y = std::numeric_limits<float>::quiet_NaN();
}
v_endif;
```

Every NaN magnitude exceeds `0x7F800000` as an integer while inf equals it and finite values fall below, so one `sfpabs` plus one integer compare separates all three.

Verified on Blackhole p100a: `pow(0, NaN)` → `0x7FC00000` on fp32, `pow(0, +inf)` → `0` preserved, `594 passed / 3 xfailed` on `test_pow.py` + `test_unary_pow.py`, no register-spill ICE.

Cost, binary `pow`, cycles/tile via tracy:

| dtype | current | with guard | delta |
|---|---:|---:|---:|
| bf16 | 3295.60 | 3519.70 | +224.10 (+6.80%) |
| fp32 | 5639.63 | 5863.22 | +223.59 (+3.96%) |

+7 SFPU instructions (144 → 151 in `trisc1.elf`, `.text` 1760 → 1788). That is 2.3× what the `v_and` refactor saved, and more expensive than the flat-`v_if` shape that preceded it. This cost is why the fix was not bundled into #53367.

Not measured on Wormhole.

### Open questions

- Is a ~4–8% regression on every binary `pow` tile acceptable to fix a masked-NaN corner? If not, is a cheaper mantissa-aware test available? An `exexp`/`exman` formulation might save 1–2 instructions, not 7.
- bf16 stays wrong regardless (inf, not NaN) until tenstorrent/tt-llk#675 is resolved, which makes the fix half-effective there. Worth deciding whether fp32-only correctness justifies the cost on both dtypes.
- The `Special Cases` doc block in both `ckernel_sfpu_binary_pow.h` copies documents `base = 0, pow < 0` and `base < 0, pow` cases but says nothing about a NaN exponent. Whatever behaviour is chosen should be written down there.

### Repro

```python
import torch, ttnn
d = ttnn.open_device(device_id=0)
shape = [1, 1, 32, 32]
z   = ttnn.from_torch(torch.zeros(shape), dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=d)
nan = ttnn.from_torch(torch.full(shape, float("nan")), dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=d)
print(ttnn.to_torch(ttnn.pow(z, nan)).flatten()[0])           # 0.0 -- expected nan
print(ttnn.to_torch(ttnn.pow(z, float("nan"))).flatten()[0])  # nan -- unary, correct
```

### Files

| file | sites |
|---|---|
| `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_pow.h` | `_sfpu_binary_power_21f_`, `_sfpu_binary_power_f32_` zero-base guards |
| `tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_pow.h` | same |

`ckernel_sfpu_rpow.h` inherits the binary kernel and needs no edit of its own. Unary (`ckernel_sfpu_unary_power.h`) is already correct: a NaN scalar fails `pow_scalar >= 0.0f` and selects the `<false>` instantiation, which fills NaN.

Follow-up to #53367 / #51468.

**Comment by iklemenskiTT** [comment]
Measured one more variant, which rules out the cheapest-looking approach.

The nested `v_if(pow < 0)` / `v_elseif(pow > 0)` shape — the form proposed in #51468 — also returns `0x00000000` for a NaN exponent, on both dtypes. `v_elseif(pow > 0.f)` is a bit-pattern test and positive quiet NaN `0x7FC00000` satisfies `> 0`, so the else-if arm fires and fills `0`. It does not leave the lane on the computed path.

All candidate guard shapes are therefore equivalent here (Blackhole p100a, fp32 binary):

| shape | `pow(0, NaN)` |
|---|---|
| pre-#53367, no guard | `0x7F800000` inf |
| flat `v_if` pair | `0x00000000` |
| nested `v_if` / `v_elseif` (#51468's proposal) | `0x00000000` |
| `v_and` (merged) | `0x00000000` |
| `v_and` + int-magnitude NaN test | `0x7FC00000` NaN |

So restructuring the guard cannot recover the louder `inf`, and cannot reach NaN either — an explicit mantissa-aware test is the only route. The +7 instruction / ~224 cycles-per-tile cost in the description stands as the price of fixing this at all.

**Comment by mcw-anasuya** [comment]
@iklemenskiTT I ran your suggested fix but it still leaves several more zero-base cases wrong. So I fixed those cases for FP32 only. The full fix costs ~5% but unrolling the FP32 reduced it and the net result is ~4% faster than the one that exists in main. Although, the same unroll cost ~7% on BF16.

**Edge cases tested for base 0:**
| Exponent | Before | After | Torch | Status |
|---|---|---|---|---|
| 0 < p < 1 | 0 | 0 | 0 | ok |
| p >= 1 | 0 | 0 | 0 | ok |
| +inf | 0 | 0 | 0 | ok |
| +0.0 | 1 | 1 | 1 | ok |
| -0.0 | NaN | 1 | 1 | **fixed** |
| p < 0 | NaN | inf | inf | **fixed** |
| -inf | NaN | inf | inf | **fixed** |
| +NaN | 0 | NaN | NaN | **fixed** |
| -NaN | NaN | NaN | NaN | ok |

**Perf:**
| variant | 32x32 | 1024x1024 | 4096x4096 |
|---|---|---|---|
| main | 10,682 ns | 141,425 ns | 2,085,624 ns |
| fixed (no unroll) | 11,159 ns (+4.46%) | 149,205 ns (+5.50%) | 2,214,019 ns (+6.16%) |
| fixed (unroll 4) | **10,450 ns (-2.17%)** | **135,091 ns (-4.48%)** | **1,982,414 ns (-4.95%)** |

Will post a PR soon

**Comment by mcw-anasuya** [comment]
FP32 fix landed on main https://github.com/tenstorrent/tt-metal/pull/55461
